### PR TITLE
fix: aws-auth cm deletes managed nodegroups entry

### PR DESCRIPTION
**Pull request by joanayma**
# PR o'clock

I think there are open issues related to this but I'm not sure.

## Description

The first apply of this module when used with `node_groups` generates a correct aws-auth file, that it's updated by AWS to add managed node groups access.
The second apply will destroy that entry that leave the managed nodes in DEGRADED state with the :
```
  # module.cluster.kubernetes_config_map.aws_auth[0] will be updated in-place
  ~ resource "kubernetes_config_map" "aws_auth" {
      ~ data        = {
          ~ "mapRoles"    = <<-EOT
                - groups:
                  - system:bootstrappers
                  - system:nodes
                  rolearn: arn:aws:iam::REDACTED:role/k8s-REDACTED20210811080122700700000009
                  username: system:node:{{EC2PrivateDNSName}}
                - groups:
                  - system:masters
                  rolearn: arn:aws:iam::REDACTED:role/path-to-another-user-role/REDACTED
                  username: admin
                - groups:
                  - system:masters
                  rolearn: arn:aws:iam::REDACTED:role/path-to-admin-role/REDACTED
                  username: admin
                - groups:
                  - system:readonly
                  rolearn: arn:aws:iam::REDACTED:role/path-to-user-role/REDACTED
                  username: admin
              - - groups:
              -   - system:bootstrappers
              -   - system:nodes
              -   rolearn: arn:aws:iam::REDACTED:role/path/to-my/k8s-iam/k8s-REDACTED20210811080122700700000009
              -   username: system:node:{{EC2PrivateDNSName}}
            EOT
            # (2 unchanged elements hidden)
        }
        id          = "kube-system/aws-auth"
        # (1 unchanged attribute hidden)

        # (1 unchanged block hidden)
    }
```

After applying the changes on this PR, the entry is kept (terraform map entry removed/added):
```
                  - system:bootstrappers
                  - system:nodes
                  rolearn: arn:aws:iam::REDACTED:role/k8s-REDACTED20210811080122700700000009
                  username: system:node:{{EC2PrivateDNSName}}
                - groups:
              +   - system:bootstrappers
              +   - system:nodes
              +   rolearn: arn:aws:iam::REDACTED:role/path/to-my/k8s-iam/k8s-REDACTED20210811080122700700000009
              +   username: system:node:{{EC2PrivateDNSName}}
              + - groups:
[...]
              - - groups:
              -   - system:bootstrappers
              -   - system:nodes
              -   rolearn: arn:aws:iam::REDACTED:role/path/to-my/k8s-iam/k8s-REDACTED20210811080122700700000009
              -   username: system:node:{{EC2PrivateDNSName}}
```

### Checklist

- [x] pre-commit hooks passed:
```
Terraform fmt............................................................Passed
Terraform docs...........................................................Failed
hookid: terraform_docs

Files were modified by this hook.

Terraform validate.......................................................Passed
Terraform validate with tflint...........................................Passed
```

- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by joanayma** [comment]
Can anyone reopen and take a look to this please?

**Comment by daroga0002** [comment]
@joanayma  could you open a issue with this problem and describing how to replicate it and what we are trying to address by this PR?

Maybe this is related to https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1595 ?

**Comment by joanayma** [comment]
Exactly. As far as I tested, the managed worker nodes will be unhealthy if the worker role in aws_auth configmap is not with the full path. In the other hand, the AWS IAM authenticatior requires it without the IAM path. Actually there's an ancient issue open about this https://github.com/kubernetes-sigs/aws-iam-authenticator/issues/268.

File: aws_auth.tf [hunk 1]
```diff
@@ -36,28 +36,41 @@ locals {
...
+            "system:nodes",
+          ],
+          role["platform"] == "windows" ? ["eks:kube-proxy-windows"] : [],
+          role["platform"] == "fargate" ? ["system:node-proxier"] : [],
+        ))
+      }
+    ],
+    [
+      # EKS managed node groups needs full role arn
+      for role in module.node_groups.aws_auth_roles :
```
Comment (joanayma) [inline aws_auth.tf:63]: Here I'm repeating the worker IAM role ARN to add it with the full IAM path.

File: aws_auth.tf [hunk 2]
```diff
@@ -78,12 +91,12 @@ resource "kubernetes_config_map" "aws_auth" {
   }
 
   data = {
-    mapRoles = yamlencode(
+    mapRoles = replace(yamlencode(
```
Comment (joanayma) [inline aws_auth.tf:94]: Here I'm removing the quotes from the formated yaml to match the diff of kubernetes_config_map.aws_auth.

File: aws_auth.tf [hunk 3]
```diff
@@ -78,12 +91,12 @@ resource "kubernetes_config_map" "aws_auth" {
...
       distinct(concat(
         local.configmap_roles,
         var.map_roles,
       ))
-    )
+    ), "\"", "")
```
Comment (joanayma) [inline aws_auth.tf:99]: Here I'm removing the quotes from the formated yaml to match the diff of kubernetes_config_map.aws_auth.

**Comment by joanayma** [comment]
It is working fine without this changes.

**Pull request closed without merging**